homeassistant/components/voip/voip.py

```diff
@@ -0,0 +1,179 @@
+from homeassistant.const import __version__
+from homeassistant.core import HomeAssistant
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class HassVoipDatagramProtocol(VoipDatagramProtocol):
+    """HA UDP server for Voice over IP (VoIP)."""
+
+    def __init__(self, hass: HomeAssistant, allow_ips: set[str]) -> None:
+        """Set up VoIP call handler."""
+        super().__init__(
+            sdp_info=SdpInfo(
+                username="homeassistant",
+                id=time.monotonic_ns(),
+                session_name="voip_hass",
+                version=__version__,
+            ),
+            protocol_factory=lambda call_info: PipelineRtpDatagramProtocol(
+                hass,
+                hass.config.language,
+            ),
+        )
+        self.allow_ips = allow_ips
+
+    def is_valid_call(self, call_info: CallInfo) -> bool:
+        """Filter calls."""
+        return call_info.caller_ip in self.allow_ips
+
+
+class PipelineRtpDatagramProtocol(RtpDatagramProtocol):
+    """Run a voice assistant pipeline in a loop for a VoIP call."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        language: str,
+    ) -> None:
+        """Set up pipeline RTP server."""
+        # STT expects 16Khz mono with 16-bit samples
+        super().__init__(rate=16000, width=2, channels=1)
+
+        self.hass = hass
+        self.language = language
+        self.pipeline: Pipeline | None = None
+
+        self._audio_queue: "asyncio.Queue[bytes]" = asyncio.Queue()
+        self._pipeline_task: asyncio.Task | None = None
+        self._conversation_id: str | None = None
+
+    def connection_made(self, transport):
+        """Server is ready."""
+        self.transport = transport
+        _LOGGER.debug(
+            "Started pipeline server on %s",
+            self.transport.get_extra_info("sockname"),
+        )
+
+    def on_chunk(self, audio_bytes: bytes) -> None:
+        """Handle raw audio chunk."""
+        if self._pipeline_task is None:
+            # Clear audio queue
+            while not self._audio_queue.empty():
+                self._audio_queue.get_nowait()
+
+            # Run pipeline until voice command finishes, then start over
+            self._pipeline_task = self.hass.async_create_background_task(
+                self._run_pipeline(),
+                "voip_pipeline_run",
+            )
+
+        self._audio_queue.put_nowait(audio_bytes)
+
+    async def _run_pipeline(
+        self,
+        pipeline_timeout: float = 30.0,
+        audio_timeout: float = 2.0,
+    ) -> None:
+        """Forward audio to pipeline STT and handle TTS."""
+        _LOGGER.debug("Starting pipeline")
+
+        async def stt_stream():
+            segmenter = VoiceCommandSegmenter()
+
+            try:
+                # Timeout if no audio comes in for a while.
+                # This means the caller hung up.
+                while chunk := await asyncio.wait_for(
```

Comment:
Prefer to use `import async_timeout` and `async with async_timeout.timeout(audio_timeout):`